File: packages/nvisy-plugin-core/src/loaders/plaintext.ts
/**
 * Plaintext loader.
 *
 * Converts `.txt` blobs into Documents by decoding the raw bytes
 * with a configurable character encoding.
 *
 * @module
 */

import { type Blob, Document, Loader } from "@nvisy/core";
import { z } from "zod";

/** Schema for plaintext loader parameters. */
export const plaintextParamsSchema = z
	.object({
		/** Character encoding of the blob data. Defaults to "utf-8". */
		encoding: z
			.enum(["utf-8", "ascii", "latin1", "utf16le"])
			.optional()
			.default("utf-8"),
	})
	.strict();

export type PlaintextParams = z.infer<typeof plaintextParamsSchema>;

/**
 * Loader that converts plaintext blobs (.txt files) into Documents.
 *
 * Reads the blob data as text using the specified encoding and
 * creates a Document with the text content.
 */
export const plaintextLoader = Loader.define<PlaintextParams>("plaintext", {
	extensions: [".txt"],
	contentTypes: ["text/plain"],
	params: plaintextParamsSchema,
	load: loadPlaintext,
});

async function* loadPlaintext(
	blob: Blob,
	params: PlaintextParams,
): AsyncGenerator<Document> {
	const content = blob.data.toString(params.encoding);
	const doc = new Document(content);
	doc.deriveFrom(blob);
	yield doc;
}

File: crates/nvisy-detect/src/loaders/plaintext.rs
use async_trait::async_trait;

use nvisy_core::datatypes::blob::Blob;
use nvisy_core::datatypes::document::Document;
use nvisy_core::errors::NvisyError;
use nvisy_core::traits::loader::{Loader, LoaderOutput};

pub struct PlaintextLoader;

#[async_trait]
impl Loader for PlaintextLoader {
    fn id(&self) -> &str {
        "plaintext"
    }

    fn extensions(&self) -> &[&str] {
        &["txt", "text"]
    }

    fn content_types(&self) -> &[&str] {
        &["text/plain"]
    }

    async fn load(
        &self,
        blob: &Blob,
        _params: &serde_json::Value,
    ) -> Result<Vec<LoaderOutput>, NvisyError> {
        let content = String::from_utf8(blob.content.to_vec()).map_err(|e| {
            NvisyError::validation(
                format!("Invalid UTF-8 in plaintext: {}", e),
                "plaintext-loader",
            )
        })?;
        let mut doc = Document::new(content);
        doc.source_format = Some("txt".to_string());
        doc.data.parent_id = Some(blob.data.id);
        Ok(vec![LoaderOutput::Document(doc)])
    }
}
